### ISSUE TITLE
server: regression in BenchmarkColBatchScan due to x-region, x-zone metrics

### ISSUE THREAD

**Issue by yuzefovich** [cockroachdb/cockroach#111148]
Here is the impact of #104585 on `BenchmarkColBatchScan` from `pkg/sql/colflow`:
```
name                        old time/op    new time/op    delta
ColBatchScan/rows=16-24       72.5µs ± 5%   104.0µs ±14%  +43.37%  (p=0.000 n=10+10)
ColBatchScan/rows=256-24       172µs ± 9%     192µs ± 1%  +11.55%  (p=0.000 n=10+8)
ColBatchScan/rows=4096-24     1.40ms ± 1%    1.42ms ± 1%   +1.78%  (p=0.000 n=8+10)
ColBatchScan/rows=65536-24    21.5ms ± 1%    20.8ms ± 1%   -3.24%  (p=0.000 n=9+8)

name                        old speed      new speed      delta
ColBatchScan/rows=16-24     3.54MB/s ± 5%  2.48MB/s ±13%  -29.82%  (p=0.000 n=10+10)
ColBatchScan/rows=256-24    23.9MB/s ± 8%  21.3MB/s ± 1%  -10.51%  (p=0.000 n=10+8)
ColBatchScan/rows=4096-24   46.8MB/s ± 1%  46.0MB/s ± 1%   -1.75%  (p=0.000 n=8+10)
ColBatchScan/rows=65536-24  48.9MB/s ± 1%  50.5MB/s ± 1%   +3.35%  (p=0.000 n=9+8)

name                        old alloc/op   new alloc/op   delta
ColBatchScan/rows=16-24       8.34kB ±13%    8.65kB ± 4%     ~     (p=0.266 n=10+8)
ColBatchScan/rows=256-24      27.9kB ± 0%    28.8kB ± 1%   +3.03%  (p=0.000 n=9+8)
ColBatchScan/rows=4096-24      195kB ±13%     192kB ± 0%     ~     (p=0.114 n=8+9)
ColBatchScan/rows=65536-24    2.19MB ± 1%    2.23MB ±16%     ~     (p=0.481 n=9+8)

name                        old allocs/op  new allocs/op  delta
ColBatchScan/rows=16-24          114 ±12%       118 ± 2%     ~     (p=0.145 n=10+8)
ColBatchScan/rows=256-24         158 ± 0%       168 ± 1%   +6.33%  (p=0.000 n=7+8)
ColBatchScan/rows=4096-24       245 ±134%       207 ± 2%     ~     (p=0.084 n=8+9)
ColBatchScan/rows=65536-24       321 ±46%     1826 ±382%     ~     (p=0.474 n=9+9)
```
This issue seems similar to #111142.

Jira issue: CRDB-31790

**Comment by yuzefovich** [comment]
Regression can also be seen on `BenchmarkKV/Scan/Native` from `pkg/sql/tests`:
```
name                          old time/op    new time/op    delta
KV/Scan/Native/rows=1-24        43.4µs ± 1%    55.5µs ± 2%  +27.92%  (p=0.000 n=9+9)
KV/Scan/Native/rows=10-24       46.2µs ± 1%    58.8µs ± 1%  +27.07%  (p=0.000 n=10+9)
KV/Scan/Native/rows=100-24      66.3µs ± 2%    79.7µs ± 2%  +20.22%  (p=0.000 n=9+9)
KV/Scan/Native/rows=1000-24      257µs ± 3%     278µs ± 1%   +7.81%  (p=0.000 n=10+9)
KV/Scan/Native/rows=10000-24    2.05ms ± 1%    2.18ms ± 2%   +6.56%  (p=0.000 n=10+10)

name                          old alloc/op   new alloc/op   delta
KV/Scan/Native/rows=1-24        9.33kB ± 0%   10.32kB ± 0%  +10.67%  (p=0.000 n=10+8)
KV/Scan/Native/rows=10-24       10.8kB ± 0%    11.8kB ± 0%   +9.25%  (p=0.000 n=9+9)
KV/Scan/Native/rows=100-24      23.7kB ± 0%    24.7kB ± 0%   +4.43%  (p=0.000 n=10+10)
KV/Scan/Native/rows=1000-24      177kB ± 0%     179kB ± 0%   +0.63%  (p=0.000 n=10+10)
KV/Scan/Native/rows=10000-24    1.55MB ± 0%    1.55MB ± 0%   +0.18%  (p=0.000 n=9+9)

name                          old allocs/op  new allocs/op  delta
KV/Scan/Native/rows=1-24          74.3 ± 1%      87.0 ± 0%  +17.09%  (p=0.000 n=10+8)
KV/Scan/Native/rows=10-24         79.0 ± 0%      92.0 ± 0%  +16.46%  (p=0.000 n=10+9)
KV/Scan/Native/rows=100-24        88.0 ± 0%     101.7 ± 1%  +15.57%  (p=0.000 n=8+10)
KV/Scan/Native/rows=1000-24        147 ± 5%       162 ± 5%  +10.39%  (p=0.000 n=10+10)
KV/Scan/Native/rows=10000-24       640 ± 5%       681 ± 4%   +6.34%  (p=0.000 n=10+9)
```

**Comment by pavelkalinnikov** [comment]
We've seen the same here https://github.com/cockroachdb/cockroach/issues/111561#issuecomment-1781020390: an error-returning path in `CompareWithLocality` func does quite a lot of allocations. Seemingly unexpected errors.

@yuzefovich In your profiles, did you see the same stack?

**Comment by yuzefovich** [comment]
I didn't look at the profiles as I only did the bisect.

**Comment by kvoli** [comment]
Will be closed on #113302 merging.

**Comment by kvoli** [comment]
Closed on https://github.com/cockroachdb/cockroach/pull/113302.
